README.md
# Getting Started with Create React App

This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).

## Available Scripts

In the project directory, you can run:

### `npm start`

Runs the app in the development mode.\
Open [http://localhost:3000](http://localhost:3000) to view it in your browser.

The page will reload when you make changes.\
You may also see any lint errors in the console.

### `npm test`

Launches the test runner in the interactive watch mode.\
See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.

### `npm run build`

Builds the app for production to the `build` folder.\
It correctly bundles React in production mode and optimizes the build for the best performance.

The build is minified and the filenames include the hashes.\
Your app is ready to be deployed!

See the section about [deployment](https://facebook.github.io/create-react-app/docs/deployment) for more information.

### `npm run eject`

**Note: this is a one-way operation. Once you `eject`, you can't go back!**

If you aren't satisfied with the build tool and configuration choices, you can `eject` at any time. This command will remove the single build dependency from your project.

Instead, it will copy all the configuration files and the transitive dependencies (webpack, Babel, ESLint, etc) right into your project so you have full control over them. All of the commands except `eject` will still work, but they will point to the copied scripts so you can tweak them. At this point you're on your own.

You don't have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn't feel obligated to use this feature. However we understand that this tool wouldn't be useful if you couldn't customize it when you are ready for it.

## Learn More

You can learn more in the [Create React App documentation](https://facebook.github.io/create-react-app/docs/getting-started).

To learn React, check out the [React documentation](https://reactjs.org/).

### Code Splitting

This section has moved here: [https://facebook.github.io/create-react-app/docs/code-splitting](https://facebook.github.io/create-react-app/docs/code-splitting)

### Analyzing the Bundle Size

This section has moved here: [https://facebook.github.io/create-react-app/docs/analyzing-the-bundle-size](https://facebook.github.io/create-react-app/docs/analyzing-the-bundle-size)

### Making a Progressive Web App

This section has moved here: [https://facebook.github.io/create-react-app/docs/making-a-progressive-web-app](https://facebook.github.io/create-react-app/docs/making-a-progressive-web-app)

### Advanced Configuration

This section has moved here: [https://facebook.github.io/create-react-app/docs/advanced-configuration](https://facebook.github.io/create-react-app/docs/advanced-configuration)

### Deployment

This section has moved here: [https://facebook.github.io/create-react-app/docs/deployment](https://facebook.github.io/create-react-app/docs/deployment)

### `npm run build` fails to minify

This section has moved here: [https://facebook.github.io/create-react-app/docs/troubleshooting#npm-run-build-fails-to-minify](https://facebook.github.io/create-react-app/docs/troubleshooting#npm-run-build-fails-to-minify)

![](https://img.shields.io/badge/build-v2.0.6-blue.svg) ![](https://img.shields.io/github/languages/code-size/EngCode/Tornado-UI-v2.svg?style=flat) ![](https://img.shields.io/github/repo-size/EngCode/Tornado-UI-v2.svg?style=flat) ![](https://img.shields.io/github/issues/EngCode/Tornado-UI-v2.svg?style=flat) ![](https://img.shields.io/github/license/EngCode/Tornado-UI-v2.svg?style=flat) ![](https://img.shields.io/badge/%40typescript-v3.9.5-blue.svg) ![](https://img.shields.io/badge/Sass-v4.13.0-blue.svg)

## Welcome to Tornado UI v2
Tornado is an Open Source Framework for Forntend and UI/UX Designers made with love from a designer to designers the Main Goal of this framework is to make web design more simple and providing the best performance and easy to control elements Builded With Sass and TypeScript for Easy Customizing.

--------------------
#### browser support
| ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari/safari_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/opera/opera_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/edge/edge_48x48.png) | ![](https://raw.githubusercontent.com/alrra/browser-logos/master/src/archive/internet-explorer_9-11/internet-explorer_9-11_48x48.png) |
| :------------: | :------------: | :------------: | :------------: | :------------: | :------------: |
| v31+ | v49+ | v10+ | v35+ | v12+ | v11+ |

--------------------
#### Documentation
looking for Documentation **Visit** [Tornado UI v2](https://tornado.phenixthemes.com/docementation/getting-start/)

--------------------
#### Before You Start
Before you start working with the framework you must know that this framework ment to be a complete work enviroment not just including the production css/js files so if you want to understand what exactly happening and use it in the right way continue reading the docementation.

#### Installation

in order to install Tornado env you can use npm or you can use Git CLI from the Following Commands.

	====== Git CLI Clone =====
	git clone https://github.com/EngCode/Tornado-UI-v2.git

	====== NPM Install ======
	npm install tornado-ui

after that you will need to install Node.JS and SASS the Ruby version for compiling the source files.

#### Quick CDN Links

	====== Tornado CSS [LTR] =====
	https://unpkg.com/tornado-ui@2.5.0/dist/css/tornado.css

	====== Tornado CSS [RTL] =====
	https://unpkg.com/tornado-ui@2.5.0/dist/css/tornado-rtl.css

	====== Tornado [ICONS] CSS =====
	https://unpkg.com/tornado-ui@2.5.0/dist/css/tornado-icons.css

	====== Tornado JS =====
	https://unpkg.com/tornado-ui@2.5.0/dist/js/tornado.min.js

--------------------
#### Upcoming Update V2.6.☯ Advanced Multimedia || 08/2020
	#-- - [new] Advanced Video Player
	#-- - [new] Advanced Audio Player
	#-- - [new] Form Range Slider Control
	#-- - [upgrade] Data Tables Advanced Options

--------------------
#### Lateast Update V2.5.☯ Super 🦸‍♂️ Tornado || 07/2020
	#01 - [new] Convert all Javascript from Babel to Typescript
	#02 - [new] React Components for all UI Elements
	#03 - [new] Seperate Icons Font CSS from Tornado CSS.
	#04 - [new] JS Event Date/Time Counter Down
	#05 - [new] Dynamic Backgrounds Lazy Loading Added
	#06 - [new] images lazyloading added
	#07 - [new] CSS Variables for Design Options
	#08 - [new] Form Control Validation In Realtime
	#09 - [new] Form UI SASS Rewrite and JS Upgrade
	#10 - [new] Form Advanced Select
	#11 - [new] Form Advanced File Uploader
	#12 - [new] Form Repeater Functionality
	#13 - [new] Form Rating Stars
	#14 - [new] Buttons new Themes Added
	#15 - [new] TNS Slider Method and UI Components
	#16 - [new] Data Tables Creation Method
	#17 - [new] Tabs TypeScript Method
	#18 - [new] Popup TypeScript Method
	#19 - [new] Images Lightbox Component and Method
	#20 - [new] Circle and Radial Progress Components
	#21 - [new] Tornado Design Options Method
	#22 - [new] Tornado Theme Switcher Method
	#23 - [doc] Documenting Typescript Methods
	#24 - [fix] JS Dynamic Page Direction for WP-Admin
	#25 - [Deprecated] Remove Lined Theme Form
	#26 - [doc] Documenting React Components
# NUB_project


changelog.md
--------------------
#### Changelog Update V2.5.☯️ Super 🦸‍♂️ Tornado || 07/2020
	#01 - [new] Convert all Javascript from Babel to Typescript
	#02 - [new] React Components for all UI Elements
	#03 - [new] Seperate Icons Font CSS from Tornado CSS.
	#04 - [new] JS Event Date/Time Counter Down
	#05 - [new] Dynamic Backgrounds Lazy Loading Added
	#06 - [new] images lazyloading added
	#07 - [new] CSS Variables for Design Options
	#08 - [new] Form Control Validation In Realtime
	#09 - [new] Form UI SASS Rewrite and JS Upgrade
	#10 - [new] Form Advanced Select
	#11 - [new] Form Advanced File Uploader
	#12 - [new] Form Repeater Functionality
	#13 - [new] Form Rating Stars
	#14 - [new] Buttons new Themes Added
	#15 - [new] TNS Slider Method and UI Components
	#16 - [new] Data Tables Creation Method
	#17 - [new] Tabs TypeScript Method
	#18 - [new] Popup TypeScript Method
	#19 - [new] Images Lightbox Component and Method
	#20 - [new] Circle and Radial Progress Components
	#21 - [new] Tornado Design Options Method
	#22 - [new] Tornado Theme Switcher Method
	#23 - [doc] Documenting Typescript Methods
	#24 - [fix] JS Dynamic Page Direction for WP-Admin
	#25 - [Deprecated] Remove Lined Theme Form
	#-- - [doc] Documenting React Components

--------------------
#### Changelog Update V2.1.0 || 12/2019
	#01 - [new] Upgrade Components into Mixins and Refactor Tornado Icons
	#02 - [new] Breadcrumb Theme Mixin
	#03 - [Doc] Documented All CSS Utlities
	#04 - [new] Adding Margin/Padding Utlities
	#05 - [new] Improve Grid CSS Utlities
	#06 - [Fix] Form Validation Message Fix
	#07 - [new] Redesign Tornado Icons [+1800 icons]
	#08 - [new] New Animations Pack Set
	#09 - [update] improve ALL JS Siblings(); Functions
	#10 - [update] Improve JS ParentsUntil(); Function
	#11 - [new] Adding JS Slide Up and Slide Down && Slide Toggle Functions.
	#12 - [Fix] Collapse Effect Fix With Slide Toggle.
	#13 - [third party] TinySlider Added [Custom CSS] [RTL Support]
	#14 - [Deprecated] Glide.jS
	#15 - [Deprecated] Helvatica Arabic Fonts
	#16 - [Deprecated] Old Tornado Icons Font
	#17 - [Fix] Allow Node Element In Live Event Listener
	#18 - [new] JS getHeight(); Function Added.
	#19 - [update] JS Accourdion // Mobile Menu // Nested Menu Dropdowns With Slide Effects
	#20 - [Fix] Form Controls With Icons Validation Style Fix.

--------------------
#### Changelog 06/2019 Components Mixins
	#01 - Documented Responsive Query Mixins.
	#02 - Adding Form Controls Themes Mixin
	#03 - Improve Responsive Menus Mixins
	#04 - Adding Extra Small Responsive Mixin
	#05 - More Coloring Helpers Added
	#06 - Adding Mobile Menu Theme Mixin
	#07 - Adding Header/Navbar Theme Mixin
	#08 - ViewPort Animations Fix

--------------------
#### Changelog 05/2019 Update
	#01 - Adding More Helpers Functions [ get All Next/Previous Siblings ].
	#02 - Upgrading CSS Backgrounds [data-src] Validation Filter.
	#03 - Adding Helper Function for Selecting Ancestors [ parentsUntil() ].
	#04 - Fixing Header icons Buttons.
	#05 - JS Realtime Assets.
	#06 - Adding Animated Numbers Counter.
	#07 - Adding ViewPort Detactor.

--------------------
#### Changelog 03/2019 Update
	#01 - Adding Simple Responsive Masonry Grid
	#02 - Convert All Pixel Unites to REM
	#03 - Adding Scrollspy Effect
	#04 - Fixing Dropdown Mechanism
	#05 - Fixing Navigation Menu / Mobile Responsive Theme
	#06 - Adding Smoth Scroll Effects
	#07 - Adding Responsive Aspect Ratio Elements
	#08 - Upgrade Utilties [ Sticky Elements | Item Remover | Data Backgrounds ]
	#09 - Javascript Shorthand Helpers
	#10 - Forms Validation With Redirect After Submit

--------------------
#### Credits
-   [William Lin](https://github.com/ganlanyuan/tiny-slider) - TinySlider Creator
-   [Contributors](https://github.com/EngCode/Tornado-UI-v2/graphs/contributors)
--------------------
#### License
the terms of the [MIT License](https://opensource.org/licenses/MIT).

src/Typescript/Base/DataTable.ts
//======> Import Tornado UI Helpers <=======//
import Tornado from './Tornado';
import Forms from './Forms';

/**======> Referance By Comment <======
 * ===> 01 - DataTable Method
 * ===> 02 - Detact Responsive Tables
 * ===> 03 - Label Translate
 * ===> 04 - Data Table Options
 * ===> 05 - Get Table Elements
 * ===> 06 - Create Labels from Data
 * ===> 07 - Create Data from Source File
 * ===> 08 - Editable Handler
 * ===> 09 - Per-Page Template
 * ===> 10 - Search Template
 * ===> 11 - Navigation Template
 * ===> 12 - Pagination Template
 * ===> 13 - Set Page Handler
 * ===> 14 - Pagination Handlers
 * ===> 15 - Create Pagination
 * ===> 16 - Per-Page Activate
 * ===> 17 - Search Handler
 * ===> 18 - Sorting Handler
*/

//=====> DataTable Method <=====//
const DataTable = (selector, options?) => {
    //====> Check if Selector or Element <====//
    typeof(selector) === 'object' ? selector = [selector] : selector = Tornado.getElements(selector);
    //====> For Each Table <====//
    selector.forEach(element => {
        var tableParent = element;
        //====> Detact Responsive Tables <====//
        if (element.parentNode.matches('.responsive-table' || '.responsive-sm-table')) tableParent = element.parentNode;
        element.classList.add('data-table-wrap');

        //====> Label Translate <====//
        const Label_Translate = (english,arabic) => {
            if (Tornado.direction('page') == 'ltr') {
                return english;
            } else {
                return arabic;
            }
        };
        
        //====> Data Table Options <====//
        options = {
            //===> Controls Options <===//
            perPage  : options?.perPage  || false, //====> Rows Per Pages
            search   : options?.search   || true,  //====> Enable Search
            sortable : options?.sortable || true,  //====> Enable Sorting
            print    : options?.print    || true, //====> Enable Printer
            //===> Data Options <===//
            data : {
                editable : options?.data?.editable || false,  //====> Enable Data Edit
                source   : options?.data?.source   || false,  //====> Data Source URL [csv, json]
                labels   : options?.data?.labels   || false,  //====> Table Head Items [array of strings]
            },
        };

        //====> Get Table Elements <====//
        var tableLabels = element.querySelectorAll('thead th'),
            tableRows   = element.querySelectorAll('tbody tr'),
            rowLength = tableRows.length;

        //====> Create Labels from Data <====//
        if (options.data.labels) {
            //====> Remove Head <====//
            element.querySelector('thead')?.remove();
            //====> Create Head Wraper <====//
            Tornado.appendIn(`<thead><tr></tr></thead>`,element);
            var headRow = element.querySelector('thead tr');
            //====> Create Labels <====//
            for (let i = 0; options.data.labels.length-1 >= i; i++) {
                var label = options.data.labels[i];
                Tornado.appendIn(`<th>${label}</th>`,headRow);
            };
            //====> get the New Labels <====//
            tableLabels = element.querySelectorAll('thead th');
        }

        //====> Create Data from Source File <====//
        if (options.data.source) {
            //====> Remove Head <====//
            element.querySelector('tbody')?.remove();
            //====> Create Head Wraper <====//
            Tornado.appendIn(`<tbody></tbody>`,element);
            var tbody = element.querySelector('tbody');
            //====> Create Labels <====//
            // for (let i = 0; options.data.labels.length-1 >= i; i++) {
            //     var label = options.data.labels[i];
            //     Tornado.appendIn(`<th>${label}</th>`,headRow);
            // };
            //====> get the New Labels <====//
            tableRows = element.querySelector('tbody tr');
        }

        //====> Editable Handler <====//
        if (options.data.editable) {
            tableRows.forEach(row => {
                row.querySelectorAll('td').forEach(column => column.setAttribute('contenteditable',true));
            });
        }

        //====> Per-Page Template <====//
        var perPageHTML = '';
        if (options.perPage) perPageHTML = `<!-- Per Page Selector -->
            <select class="form-control small perpage">
                <option value="${options.perPage}">${options.perPage}</option>
                ${options.perPage*2 < rowLength ? `<option value="${options.perPage*2}">${options.perPage*2}</option>` : ''};
                ${options.perPage*4 < rowLength ? `<option value="${options.perPage*4}">${options.perPage*4}</option>` : ''};
                ${options.perPage*10 < rowLength ? `<option value="${options.perPage*10}">${options.perPage*10}</option>` : ''};
                ${options.perPage*20 < rowLength ? `<option value="${options.perPage*20}">${options.perPage*20}</option>` : ''};
                ${options.perPage*25 < rowLength ? `<option value="${options.perPage*25}">${options.perPage*25}</option>` : ''};
                ${options.perPage*50 < rowLength ? `<option value="${options.perPage*50}">${options.perPage*50}</option>` : ''};
                ${options.perPage*100 < rowLength ? `<option value="${options.perPage*100}">${options.perPage*100}</option>` : ''};
                ${`<option value="${rowLength}">${rowLength}</option>`};
            </select>
            <label class="backspace">${Label_Translate('Per Page', 'فى كل صفحة')}</label>
        `;

        //====> Search Template <====//
        var searchHTML = '';
        if (options.search) searchHTML = `
            <div class="control-icon small floating-end ti-search">
                <input type="search" class="table-search form-control small" placeholder="${Label_Translate('Search in Table', 'البحث فى الجدول')}">
            </div>
        `;

        //====> Navigation Template <====//
        var tableNav = `<div class="data-table-nav flexbox align-between align-center-y">
            ${options.perPage ? 
                `<!-- Action Area -->
                <div class="table-filters col-auto">
                    ${perPageHTML}
                </div>` : ''
            }

            ${options.print ?             
                `<!-- Printer Button -->
                <div class="table-filters">
                    <a href="javascript:void(0);" class="btn small default btn-icon ti-print print-table">طباعة</a>
                </div>
                <!-- // Printer Button -->` : ''
            }
            ${options.search ?             
                `<!-- Seach -->
                <div class="table-filters">
                    ${searchHTML}
                </div>
                <!-- // Seach -->` : ''
            }
        </div>`;

        //====> Insert Navigation <====//
        Tornado.insBefore(tableNav, tableParent);

        //====> Pagination Template <====//
        var paginationTemplate = `<div class="data-table-pagi flexbox align-between align-center-y">
            <!-- Display Info -->
            <p>${Label_Translate(
                `Showing from <span class="start-count">1</span> to <span class="end-count">${tableRows.length}</span> out of ${tableRows.length}`,
                `اظهار <span class="start-count">1</span> الي <span class="end-count">${tableRows.length}</span> من مجموع ${tableRows.length}`
            )}</p>
            <!-- Pagination -->
            <ul class="pagination small"></ul>
        </div>`;

        //====> Insert Pagination <====//
        Tornado.insertAfter(paginationTemplate, tableParent);

        //====> get Navigation and Pagination <====//
        var navigation = Tornado.getPrevSibling({element:tableParent, filter:'.data-table-nav'});
        var pagination = Tornado.getNextSibling({element:tableParent, filter:'.data-table-pagi'});

        //====> Set Page Handler <====//
        const setPage = (rows, start, end) => {
            //====> Clear Current Items <====//
            element.querySelectorAll('tbody tr')?.forEach(del => del.remove());
            //====> Set Pages <====//
            for(let i = start; end >= i; i++) {
                if(rows[i]) Tornado.appendIn(rows[i], element.querySelector('tbody'))
            };
            //====> Display Information <====//
            pagination.querySelector('.start-count').textContent = start+1;
            pagination.querySelector('.end-count').textContent = end+1;
        };

        //====> Pagination Handlers <====//
        const paginationHandlers = () => {
            //===> Set New Page <===//
            const setNewPage = item => {
                //====> Positions <====//
                var thisPage   = parseInt(item.textContent),
                    perPageNumber = navigation.querySelector('.perpage').value;

                //====> Activate <====//
                item.parentNode.querySelector('.active')?.classList.remove('active')
                item.classList.add('active');

                //====> Items Caluclation <====//
                var startPoint = thisPage == 1 ? 0 : thisPage*perPageNumber-perPageNumber,
                    endPoint   = thisPage*perPageNumber-1;

                //====> Set First Page <====//
                setPage(tableRows, startPoint, endPoint);
            };

            //===> Get Pagination Items <===//
            pagination.querySelectorAll('.pagination li:not(.next-page):not(.prev-page)').forEach(item => {
                //===> On Item Click <===//
                item.addEventListener('click', event => {
                    event.preventDefault();
                    setNewPage(event.target);
                });
            });

            //===> Next Page <===//
            pagination.querySelector('.pagination .next-page')?.addEventListener('click', event => {
                event.preventDefault();
                //====> Positions <====//
                var item = Tornado.getNextSibling({element:pagination.querySelector('li.active'), filter:'li:not(.prev-page):not(.next-page)'});
                if(item) setNewPage(item);
            });

            //===> Prev Page <===//
            pagination.querySelector('.pagination .prev-page')?.addEventListener('click', event => {
                event.preventDefault();
                //====> Positions <====//
                var item = Tornado.getPrevSibling({element:pagination.querySelector('li.active'), filter:'li:not(.prev-page):not(.next-page)'});
                if (item) setNewPage(item);
            });
        };

        //====> Create Pagination <====//
        const createPagination = (rows, start, end) => {
            //====> Clear Current Items <====//
            element.querySelectorAll('tbody tr')?.forEach(del => del.remove());

            //====> Select Pagination List <====//
            var paginationList = pagination.querySelector('.pagination'),
                pagesNumber    = Math.ceil(rows.length/options.perPage);

            //===> Reset Pagination <===//
            paginationList.innerHTML = null;

            //====> Create Pagination Items <====//
            for(let i = 0; pagesNumber-1 >= i; i++) Tornado.appendIn(`<li class="${i==0?'active':''}">${i+1}</li>`, paginationList);

            //====> Create Next / Prev Buttons <====//
            if (pagesNumber > 1) {
                Tornado.insBefore(`<li class="prev-page ti-arrow-${Tornado.direction('start')}-c"></li>`, paginationList.querySelector('li:first-child'));
                Tornado.appendIn(`<li class="next-page ti-arrow-${Tornado.direction('end')}-c"></li>`, paginationList);
            }

            //====> Set First Page <====//
            setPage(rows,start,end);

            //====> Activate Pagination <====//
            paginationHandlers();
        };

        //====> Per-Page Activate <====//
        if(options.perPage) {
            //====> Create Pagination <====//
            createPagination(tableRows, 0, options.perPage-1);

            //====> Per-Page Handler <====//
            var perPageControl = navigation.querySelector('.perpage');
            Forms.advancedSelect(perPageControl);

            perPageControl.addEventListener('change', event => {
                options.perPage = event.target.value;
                //====> Create Pagination <====//
                createPagination(tableRows, 0, options.perPage-1);
            });
        }

        //====> Search Handler <====//
        if(options.search) {
            navigation.querySelector('.table-search')?.addEventListener('keyup', event => {
                //====> Get Search Data <====//
                var currentValue = event.target.value.toString().toLowerCase(),
                    keyboard_key = event.keyCode || event.charCode,
                    filteredData = [];
    
                //====> Search Process <====//
                if(keyboard_key == 8 && currentValue == '' || null) {
                    //====> if its Empty Return Original Data <====//
                    filteredData = tableRows;
                } else {
                    //=====> Loop Through Options <=====//
                    tableRows.forEach((item,index) => {
                        //=====> Search and Options Data <=====//
                        var labelValue = item.textContent.toString().toLowerCase();
                        //=====> Show / Hide Options <=====//
                        if (labelValue.includes(currentValue)) filteredData.push(item);
                    });
                }
    
                //====> Search Resault <====//
                if (options.perPage && filteredData.length > 0) createPagination(filteredData, 0, options.perPage-1);
                if (!options.perPage) setPage(filteredData, 0, filteredData.length-1);
            });
        }

        //====> Sorting Handler <====//
        if(options.sortable) {
            tableLabels.forEach((label, tdIndex) => {
                //====> Set Sort Icons <====//
                label.classList.add('sort-table');
                Tornado.appendIn(`<span class="sort-icon ti-arrow-down-c"></span>`, label);

                //====> Handler <====//
                label.addEventListener('click', event => {
                    event.preventDefault();
                    //====> Deactive Handler <====//
                    const deactiveOther = () => {
                        //====> Deactivte Other <====//
                        Tornado.getSiblings({element:label, filter:'.active'})?.forEach(sibling => {
                            sibling.classList.remove('active');
                            sibling.classList.remove('asc-sort');
                            sibling.classList.remove('desc-sort');
                        });
                    }

                    //====> Sort Detactor <====//
                    if(!label.classList.contains('asc-sort')) {
                        var sortBy = 'asc'
                        //====> Deactivte Other <====//
                        deactiveOther();
                        //====> Activate Item <====//
                        label.classList.add('active');
                        label.classList.add('asc-sort');
                        label.classList.remove('desc-sort');
                    } else {
                        var sortBy = 'desc';
                        //====> Deactivte Other <====//
                        deactiveOther();
                        //====> Activate Item <====//
                        label.classList.add('active');
                        label.classList.add('desc-sort');
                        label.classList.remove('asc-sort');
                    }

                    //====> Clear Current Items <====//
                    if(options.perPage) {
                        var tableBody = element.querySelector('tbody');
                        tableBody.querySelectorAll('tr')?.forEach(del => del.remove());
                        //====> Show All <====//
                        if (tableBody) tableRows.forEach(row => tableBody.appendChild(row));
                    }

                    //====> Rows Sorting <====//
                    element.querySelectorAll('tbody tr')?.forEach((row, index) => {
                        var rowSiblings = Tornado.getSiblings({element:row,filter:'tr'}),
                            currentValue = row.children[tdIndex].textContent;
                        //====> Compare Row to its Sibling <====//
                        rowSiblings.forEach(sibling => {
                            var siblingValue = sibling.children[tdIndex].textContent;
                            
                            //====> Compare as Number <====//
                            if (!isNaN(currentValue)) {
                                currentValue = parseFloat(currentValue);
                                siblingValue = parseFloat(siblingValue);
                            } 
                            
                            //====> Compare as String <====//
                            if (isNaN(currentValue)) {
                                currentValue = currentValue.toString();
                                siblingValue = siblingValue.toString();
                            }

                            //====> ASC Sorting <====//
                            if(sortBy == 'asc' && currentValue >= siblingValue) Tornado.insertAfter(row, sibling);
                            //====> DESC Sorting <====//
                            if(sortBy == 'desc' && currentValue <= siblingValue) Tornado.insertAfter(row, sibling);
                        });
                    });

                    //====> new Sorted Rows <====//
                    tableRows = element.querySelectorAll('tbody tr');
                    //====> Create Pagination <====//
                    if (options.perPage && tableRows.length > 0) createPagination(tableRows, 0, options.perPage-1);
                    if (!options.perPage) setPage(tableRows, 0, tableRows.length-1);
                });
            });
        }
    });
}

export default DataTable;
